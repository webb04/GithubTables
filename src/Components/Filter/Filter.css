.Filter-Wrapper {
    display: inline-block;
}
.Filter {
    display: inline-block;
    padding: 10px 20px;
    border: solid 1px #dedede;
    margin-bottom: 10px;
    user-select: none;
}

.spacer {
    margin-right: 80px;
}

.dropdown {
    position: absolute;
    background: white;
    z-index: 10;
    box-shadow: 0 3px 12px rgba(27,31,35,0.15);
    max-height: 250px;
    overflow: scroll;
    border: solid 1px #dedede;
}

.item {
    padding: 8px 8px 8px 30px;
    border-top: solid 1px lightgrey;
}

.selected {
    padding: 8px;
    text-align: center;
    font-weight: bold;
}

.dropdown input {
    margin: 10px;
    width: calc(100% - 20px);
    padding: 4px;
    box-sizing: border-box;
    font-size: 16px;
}

.clear {
    text-align: center;
    padding: 8px 0;
    color: #DA131399;
}

.clear:hover, .dropdown:hover, .Filter:hover {
    cursor: pointer;
}