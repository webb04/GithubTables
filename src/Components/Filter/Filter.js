import React from 'react';
import PropTypes from 'prop-types';

import './Filter.css';

class Filter extends React.Component {
    constructor() {
        super();
        this.state = {
          open: false,
          searchTerm: ''
        };
        this.onTextInput = this.onTextInput.bind(this);
        this.handleClickOutside = this.handleClickOutside.bind(this);
      }

    componentDidMount() {
        document.addEventListener('click', this.handleClickOutside);
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.handleClickOutside);
    }

    handleClickOutside(e) {
        if (this.node && !this.node.contains(e.target)) {
            this.setState({ open: false})
        }
    }

    onTextInput({ target }) {
        this.setState({ searchTerm: target.value });
    }

    render() {
        const { label, onSelectLabel, onSelectAuthor, options, selectedLabelFilter, selectedAuthorFilter } = this.props;
        const { open, searchTerm } = this.state;
        
        const filteredOptions = searchTerm 
            ? options.filter(option => option.toLowerCase().indexOf(searchTerm) >= 0)
            : options;

        const onChange = label === 'Authors' ? onSelectAuthor : onSelectLabel;
        const selected = label === 'Authors' ? selectedAuthorFilter : selectedLabelFilter;

        return (
                <div className="Filter-Wrapper" ref={node => this.node = node}>
                    <div className="Filter" onClick={() => this.setState({ open: !open })}>
                        {`Filter ${label}`}
                    </div>
                    {
                        open ?
                            <div className="dropdown">
                                <input type="text" onChange={this.onTextInput} value={searchTerm}/>
                                {
                                    selected 
                                        ? <div>
                                            <div className="selected">{selected}</div>
                                            <div className="clear" onClick={() => onChange(null)}>Clear</div>
                                        </div>
                                        : null
                                }
                                {
                                    filteredOptions.map((option, index) => {
                                        return <div key={`filter-${index}`} onClick={() =>  onChange(option)} className="item">{ option }</div>
                                    })
                                }
                            </div>
                            : null
                    }
                </div>
        )
    }
}

Filter.defaultProps = {
    label: 'Authors',
    onSelectLabel: () => {},
    onSelectAuthor: () => {},
    selectedLabelFilter: '',
    selectedAuthorFilter: ''
}

Filter.propTypes = {
    label: PropTypes.oneOf(['Authors', 'Labels']),
    onSelectLabel: PropTypes.func,
    onSelectAuthor: PropTypes.func,
    options: PropTypes.array.isRequired,
    selectedLabelFilter: PropTypes.string,
    selectedAuthorFilter: PropTypes.string
}

export default Filter;