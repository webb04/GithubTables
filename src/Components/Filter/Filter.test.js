import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Filter from './Filter';
import { shallow, mount } from 'enzyme';

it('Renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Filter options={['one', 'two']} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('Renders correctly', () => {
    const tree = renderer
        .create(<Filter options={['one', 'two']} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('Opens on click', () => {
    const Component = mount(<Filter options={['one', 'two']}/>);
    expect(Component.find('.dropdown')).toHaveLength(0);
    Component.find('.Filter').simulate('click');
    expect(Component.state().open).toBe(true);
    expect(Component.find('.dropdown')).toHaveLength(1);
    expect(Component.find('.item')).toHaveLength(2);
});

it('Displays all options', () => {
    const Component = mount(<div>
        <Filter id="one" options={['one', 'two']}/>
        <Filter id="two" options={['three', 'four']}/>
        </div>);

    Component.find('#one').find('.Filter').simulate('click');
    expect(Component.contains('one')).toBe(true);
    expect(Component.contains('two')).toBe(true);
    expect(Component.contains('three')).toBe(false);
    expect(Component.contains('four')).toBe(false);

    Component.find('#two').find('.Filter').simulate('click');
    expect(Component.contains('three')).toBe(true);
    expect(Component.contains('four')).toBe(true);
});

it('Filters by search term', () => {
    const Component = mount(<Filter options={['one', 'two']}/>);
    
    Component.find('.Filter').simulate('click');
    expect(Component.contains('one')).toBe(true);
    expect(Component.contains('two')).toBe(true);
    
    Component.setState({ searchTerm: 'one' });
    expect(Component.contains('one')).toBe(true);
    expect(Component.contains('two')).toBe(false);
    
    Component.setState({searchTerm: 'eight'});
    expect(Component.find('.item')).toHaveLength(0);
});

it('Clears on click', () => {
    const Component = mount(<Filter 
        options={['one', 'two']}
        label="Authors"
        selectedAuthorFilter="Fred"
    />);

    Component.find('.Filter').simulate('click');
    expect(Component.find('.clear')).toHaveLength(1);
    expect(Component.find('.selected')).toHaveLength(1);

    Component.find('.clear').simulate('click')
    Component.setProps({
        options: ['one', 'two'],
        label: "Authors",
        selectedAuthorFilter: null
    })
    expect(Component.find('.clear')).toHaveLength(0);
    expect(Component.find('.selected')).toHaveLength(0);
});

it('Renders label', () => {
    const Component = mount(<Filter options={['five']} label="Labels"/>);
    expect(Component.contains('Filter Labels')).toBe(true);
});