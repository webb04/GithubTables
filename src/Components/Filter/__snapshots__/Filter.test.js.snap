// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="Filter-Wrapper"
>
  <div
    className="Filter"
    onClick={[Function]}
  >
    Filter Authors
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="Filter-Wrapper"
>
  <div
    className="Filter"
    onClick={[Function]}
  >
    Filter Authors
  </div>
</div>
`;
